File: bin/check-links.js
#!/usr/bin/env node
/**
 * Headless Chrome Link Checker
 *
 * Uses Puppeteer to check URLs with a real browser, bypassing bot protection.
 *
 * Usage:
 *   node check-links.js [file.md]     # Check URLs in a markdown file
 *   node check-links.js --url URL     # Check a single URL
 *   node check-links.js --stdin       # Read URLs from stdin (one per line)
 *
 * Output:
 *   JSON report to stdout with status for each URL
 */

import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

// Configuration
const CONFIG = {
  concurrency: 5,           // Parallel browser pages
  timeout: 15000,           // Page load timeout (ms)
  userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  headless: true,
  retries: 1,
};

// Extract URLs from markdown content
function extractUrls(content) {
  const urlPattern = /https?:\/\/[^\s\)>\]"']+/g;
  const matches = content.match(urlPattern) || [];
  // Deduplicate and clean
  return [...new Set(matches.map(url => url.replace(/[,.\)]+$/, '')))];
}

// Check a single URL with Puppeteer
async function checkUrl(page, url) {
  const result = {
    url,
    status: null,
    statusText: null,
    title: null,
    error: null,
    time: null,
  };

  const start = Date.now();

  try {
    const response = await page.goto(url, {
      waitUntil: 'domcontentloaded',
      timeout: CONFIG.timeout,
    });

    result.status = response?.status() || 0;
    result.statusText = response?.statusText() || 'Unknown';
    result.title = await page.title().catch(() => null);
    result.time = Date.now() - start;

  } catch (err) {
    result.error = err.message;
    result.time = Date.now() - start;

    // Try to extract status from error message
    if (err.message.includes('net::ERR_NAME_NOT_RESOLVED')) {
      result.status = 0;
      result.statusText = 'DNS_FAILED';
    } else if (err.message.includes('net::ERR_CONNECTION_REFUSED')) {
      result.status = 0;
      result.statusText = 'CONNECTION_REFUSED';
    } else if (err.message.includes('net::ERR_CONNECTION_TIMED_OUT')) {
      result.status = 0;
      result.statusText = 'TIMEOUT';
    } else if (err.message.includes('Navigation timeout')) {
      result.status = 0;
      result.statusText = 'NAV_TIMEOUT';
    }
  }

  return result;
}

// Process URLs with concurrency limit
async function processUrls(urls, browser) {
  const results = [];
  const queue = [...urls];
  const workers = [];

  for (let i = 0; i < CONFIG.concurrency; i++) {
    workers.push((async () => {
      const page = await browser.newPage();
      await page.setUserAgent(CONFIG.userAgent);
      await page.setViewport({ width: 1280, height: 720 });

      // Block unnecessary resources for speed
      await page.setRequestInterception(true);
      page.on('request', (req) => {
        const type = req.resourceType();
        if (['image', 'stylesheet', 'font', 'media'].includes(type)) {
          req.abort();
        } else {
          req.continue();
        }
      });

      while (queue.length > 0) {
        const url = queue.shift();
        if (!url) break;

        process.stderr.write(`Checking: ${url.substring(0, 60)}...\n`);

        let result = await checkUrl(page, url);

        // Retry once if failed
        if (result.status === 0 && CONFIG.retries > 0) {
          await new Promise(r => setTimeout(r, 1000));
          result = await checkUrl(page, url);
        }

        results.push(result);
      }

      await page.close();
    })());
  }

  await Promise.all(workers);
  return results;
}

// Generate summary report
function generateReport(results) {
  const summary = {
    total: results.length,
    ok: 0,
    redirects: 0,
    clientErrors: 0,
    serverErrors: 0,
    blocked: 0,
    failed: 0,
  };

  const byStatus = {};

  for (const r of results) {
    const status = r.status || 0;
    byStatus[status] = (byStatus[status] || 0) + 1;

    if (status >= 200 && status < 300) summary.ok++;
    else if (status >= 300 && status < 400) summary.redirects++;
    else if (status === 403) summary.blocked++;
    else if (status >= 400 && status < 500) summary.clientErrors++;
    else if (status >= 500) summary.serverErrors++;
    else summary.failed++;
  }

  return {
    summary,
    byStatus,
    results: results.sort((a, b) => (a.status || 999) - (b.status || 999)),
    timestamp: new Date().toISOString(),
  };
}

// Main
async function main() {
  const args = process.argv.slice(2);
  let urls = [];

  // Parse arguments
  if (args.includes('--url')) {
    const idx = args.indexOf('--url');
    urls = [args[idx + 1]];
  } else if (args.includes('--stdin')) {
    const input = fs.readFileSync(0, 'utf-8');
    urls = input.split('\n').filter(u => u.trim().startsWith('http'));
  } else if (args.length > 0 && fs.existsSync(args[0])) {
    const content = fs.readFileSync(args[0], 'utf-8');
    urls = extractUrls(content);
  } else {
    console.error('Usage:');
    console.error('  node check-links.js <file.md>      Check URLs in markdown file');
    console.error('  node check-links.js --url <URL>    Check single URL');
    console.error('  node check-links.js --stdin        Read URLs from stdin');
    process.exit(1);
  }

  if (urls.length === 0) {
    console.error('No URLs found.');
    process.exit(1);
  }

  process.stderr.write(`Found ${urls.length} URLs to check\n\n`);

  // Launch browser
  const browser = await puppeteer.launch({
    headless: CONFIG.headless,
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--disable-gpu',
    ],
  });

  try {
    const results = await processUrls(urls, browser);
    const report = generateReport(results);

    // Output JSON report
    console.log(JSON.stringify(report, null, 2));

    // Print summary to stderr
    process.stderr.write('\n--- SUMMARY ---\n');
    process.stderr.write(`Total:    ${report.summary.total}\n`);
    process.stderr.write(`OK (2xx): ${report.summary.ok}\n`);
    process.stderr.write(`Blocked:  ${report.summary.blocked}\n`);
    process.stderr.write(`Errors:   ${report.summary.clientErrors + report.summary.serverErrors}\n`);
    process.stderr.write(`Failed:   ${report.summary.failed}\n`);

  } finally {
    await browser.close();
  }
}

main().catch(err => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});

File: src/check_links.rs
//! check-links command: Check URL health
//!
//! LLM-optimized output - JSON compact only.

use crate::browser::BrowserPool;
use crate::extract::extract_urls;
use anyhow::{Context, Result};
use clap::Args;
use serde::Serialize;
use std::io::{self, BufRead};
use tokio::fs;

#[derive(Args)]
pub struct CheckLinksArgs {
    /// Markdown file to check URLs from
    #[arg(value_name = "FILE")]
    file: Option<String>,

    /// Check a single URL
    #[arg(long)]
    url: Option<String>,

    /// Read URLs from stdin (one per line)
    #[arg(long)]
    stdin: bool,

    /// Number of parallel browser tabs (1-20)
    #[arg(short, long, default_value = "5", value_parser = clap::value_parser!(u8).range(1..=20))]
    concurrency: u8,

    /// Timeout per URL in milliseconds
    #[arg(long, default_value = "15000")]
    timeout: u64,

    /// Number of retries on failure
    #[arg(long, default_value = "1")]
    retries: u8,
}

/// Configuration for check-links
pub struct CheckLinksConfig {
    pub concurrency: usize,
    pub timeout_ms: u64,
    pub retries: u8,
}

/// Result for a single link check (compact)
#[derive(Debug, Serialize)]
pub struct LinkResult {
    pub url: String,
    pub status: u16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub redirect_to: Option<String>,
}

/// Full report (compact)
#[derive(Debug, Serialize)]
pub struct LinkReport {
    pub ok: usize,
    pub failed: usize,
    pub results: Vec<LinkResult>,
}

/// Run the check-links command
pub async fn run_check_links(args: CheckLinksArgs) -> Result<()> {
    let urls = get_urls(&args).await?;

    if urls.is_empty() {
        eprintln!("No URLs found.");
        std::process::exit(1);
    }

    eprintln!(
        "Checking {} URLs ({} parallel)...",
        urls.len(),
        args.concurrency
    );

    let config = CheckLinksConfig {
        concurrency: args.concurrency as usize,
        timeout_ms: args.timeout,
        retries: args.retries,
    };

    let report = check_links(&urls, &config).await?;

    // Output compact JSON to stdout
    println!("{}", serde_json::to_string(&report)?);

    eprintln!("Done: {}/{} OK", report.ok, report.ok + report.failed);

    Ok(())
}

/// Get URLs from file, --url, or stdin
async fn get_urls(args: &CheckLinksArgs) -> Result<Vec<String>> {
    if let Some(url) = &args.url {
        return Ok(vec![url.clone()]);
    }

    if args.stdin {
        let stdin = io::stdin();
        let urls: Vec<String> = stdin
            .lock()
            .lines()
            .map_while(Result::ok)
            .filter(|line| line.starts_with("http"))
            .collect();
        return Ok(urls);
    }

    if let Some(file) = &args.file {
        let content = fs::read_to_string(file)
            .await
            .with_context(|| format!("Failed to read file: {}", file))?;
        return Ok(extract_urls(&content));
    }

    eprintln!("Usage:");
    eprintln!("  ref-tools check-links <file.md>    Check URLs in markdown file");
    eprintln!("  ref-tools check-links --url <URL>  Check single URL");
    eprintln!("  ref-tools check-links --stdin      Read URLs from stdin");
    std::process::exit(1);
}

/// Check multiple links and generate report
pub async fn check_links(urls: &[String], config: &CheckLinksConfig) -> Result<LinkReport> {
    let pool = BrowserPool::new(config.concurrency).await?;
    let mut results = Vec::with_capacity(urls.len());
    let mut ok_count = 0;
    let mut failed_count = 0;

    for url in urls {
        eprintln!("  -> {}", truncate(url, 60));

        let page = pool.new_page().await?;
        let mut result = page.goto(url, config.timeout_ms).await?;

        // Retry on failure
        if result.status == 0 && config.retries > 0 {
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
            result = page.goto(url, config.timeout_ms).await?;
        }

        // Determine if redirect (check final URL)
        let redirect_to = if result.status >= 200 && result.status < 400 {
            if let Some(final_url) = page.current_url().await {
                // Check if different domain
                let orig_host = url::Url::parse(url)
                    .ok()
                    .and_then(|u| u.host_str().map(String::from));
                let final_host = url::Url::parse(&final_url)
                    .ok()
                    .and_then(|u| u.host_str().map(String::from));

                if let (Some(orig), Some(fin)) = (orig_host, final_host) {
                    let orig_norm = orig.trim_start_matches("www.");
                    let fin_norm = fin.trim_start_matches("www.");
                    if orig_norm != fin_norm {
                        Some(final_url)
                    } else {
                        None
                    }
                } else {
                    None
                }
            } else {
                None
            }
        } else {
            None
        };

        let is_ok = result.status >= 200 && result.status < 400 && redirect_to.is_none();
        if is_ok {
            ok_count += 1;
        } else {
            failed_count += 1;
        }

        results.push(LinkResult {
            url: url.clone(),
            status: result.status,
            error: result.error,
            redirect_to,
        });
    }

    pool.close().await?;

    Ok(LinkReport {
        ok: ok_count,
        failed: failed_count,
        results,
    })
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        s.to_string()
    } else {
        format!("{}...", &s[..max - 3])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_truncate() {
        assert_eq!(truncate("short", 10), "short");
        assert_eq!(truncate("this is a very long string", 10), "this is...");
    }
}
